libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -0,0 +1,232 @@
+{
+	Whiskers t(R"(
+		object "<CreationObject>" {
+			code {
+				<memoryInit>
+				<constructor>
+				<deploy>
+				<functions>
+			}
+			object "<RuntimeObject>" {
+				code {
+					<memoryInit>
+					<dispatch>
+					<runtimeFunctions>
+				}
+			}
+		}
+	)");
+
+	resetContext();
+	t("CreationObject", creationObjectName(_contract));
+	t("memoryInit", memoryInit());
+	t("constructor", _contract.constructor() ? constructorCode(*_contract.constructor()) : "");
+	t("deploy", deployCode(_contract));
+	t("functions", m_context.functionCollector()->requestedFunctions());
+
+	resetContext();
+	t("RuntimeObject", runtimeObjectName(_contract));
+	t("dispatch", dispatchRoutine(_contract));
+	t("runtimeFunctions", m_context.functionCollector()->requestedFunctions());
+	return t.render();
+}
+
+string IRGenerator::generateIRFunction(FunctionDefinition const& _function)
+{
+	string functionName = "fun_" + to_string(_function.id()) + "_" + _function.name();
+	return m_context.functionCollector()->createFunction(functionName, [&]() {
+		Whiskers t("function <functionName>(<params>) <returns> {}");
+		t("functionName", functionName);
+		string params;
+		for (auto const& varDecl: _function.parameters())
+			params += (params.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl);
+		t("params", params);
+		string retParams;
+		for (auto const& varDecl: _function.returnParameters())
+			retParams += (retParams.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl);
+		t("returns", retParams.empty() ? "" : " -> " + retParams);
+		return t.render();
+	});
+}
+
+string IRGenerator::constructorCode(FunctionDefinition const& _constructor)
+{
+	string out;
+	if (!_constructor.isPayable())
+		out = callValueCheck();
+
+	// TODO
+
+	return out;
+}
+
+string IRGenerator::deployCode(ContractDefinition const& _contract)
+{
+	Whiskers t(R"X(
+		codecopy(0, dataoffset("<object>"), datasize("<object>"))
+		return(0, datasize("<object>"))
+	)X");
+	t("object", runtimeObjectName(_contract));
+	return t.render();
+}
+
+string IRGenerator::callValueCheck()
+{
+	return "if callvalue() { revert(0, 0) }";
+}
+
+string IRGenerator::creationObjectName(ContractDefinition const& _contract)
+{
+	return _contract.name() + "_" + to_string(_contract.id());
+}
+
+string IRGenerator::runtimeObjectName(ContractDefinition const& _contract)
+{
+	return _contract.name() + "_" + to_string(_contract.id()) + "_deployed";
+}
+
+string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
+{
+	Whiskers t(R"X(
+		if iszero(lt(calldatasize(), 4))
+		{
+			let selector := <shr224>(calldataload(0))
+			switch selector
+			<#cases>
+			case <functionSelector>
+			{
+				// <functionName>
+				<callValueCheck>
+				<assignToParams> <abiDecode>(4, calldatasize())
+				<assignToRetParams> <function>(<params>)
+				let memPos := <allocate>(0)
+				let memEnd := <abiEncode>(memPos <comma> <retParams>)
+				return(memPos, sub(memEnd, memPos))
+			}
+			</cases>
+			default {}
```

Comment:
Why is this not the same as the fallback?

Reply:
It is... this doesn't return, but will instead continue with ``<fallback>`` below, won't it?

Reply:
Yep.

Reply:
oh, right